README.md
# Flock MySQL

Flock MySQL is a Flock plugin for MySQL.

## Install

```
npm install launchfort/flock-mysql
```

## Usage

```js
// .flockrc.js
const { DefaultMigrator, NodeModuleMigrationProvider } = require('@launchfort/flock')
const { DataAccessProvider, TemplateProvider } = require('@launchfort/flock-mysql')

const migrationDir = 'migrations'
const migrationTableName = 'migration'
const connectionOptions = {
  host: 'localhost',
  database: process.env.MYSQL_DATABASE,
  user: process.env.MYSQL_USER,
  password: process.env.MYSQL_PASSWORD
}
const dap = new DataAccessProvider({ migrationTableName, connectionOptions })
const mp = new NodeModuleMigrationProvider({ migrationDir })

exports.migrator = new DefaultMigrator(mp, dap)
exports.migrationDir = migrationDir
exports.templateProvider = new TemplateProvider()

```

## Migrations

When writing migrations that use `flock-mysql` then the `QueryInterface#query`
method signature is identical to that of the [pg](https://npmjs.org/pg)'s [Client#query](https://node-postgres.com/features/queries) method.

Example:
```js
exports.up = queryInterface => {
  const sql = 'SELECT * FROM user WHERE age = ?'
  const values = [ 1 ]
  return queryInterface.query({ sql, values })
}
```

The `QueryInterface#query` method accepts a query object with the following
shape:

```ts
{
  sql: string,
  values?: any[],
  timeout?: number
}
```

## API

Flock mysql exports implementations of Flock's `DataAccessProvider` and `TemplateProvider`
as `DataAccessProvider` and `TemplateProvider` classes.

The `DataAccessProvider` class will connect to your MySQL DB by reading
the properties from `connectionOptions`. This object is the same options object
accepted by the [mysql](https://npmjs.org/mysql) module's [createConnection](https://www.npmjs.com/package/mysql#connection-options) function.

*NOTE: The only difference with the connection options is that `flock-mysql`
defaults the timezone option to `Z` instead of `local`.*

```js
class DataAccessProvider implements Flock.DataAccessProvider {
  constructor ({
    migrationTableName = 'migration',
    acquireLock = true,
    connectionOptions = /* REQUIRED */ } = {})
}
```

Additionally, by default the `DataAccessProvider` will attempt to acquire an
application lock immediately after connecting to the database. This behaviour
can be overridden by setting the `acquireLock` option to `false`. Acquiring a
lock helps to prevent concurrent migrations from occuring.

See: https://devcenter.heroku.com/articles/release-phase


CHANGELOG.md
# 1.0.0

**Major**

- Reach stabiliyt
- Change scope from @gradealabs to @launchfort

**Path**

- Update dependencies

# 0.0.5

**Patch**

- Resolve issue with rolling back without first migrating (i.e. the migration
  table did not yet exist).

# 0.0.4

**Patch**

- Fix author email address in `package.json`.
- Remove `tsconfig-paths` from `package.json`.
- Add `.eslintignore` to ignore the `lib` folder.

# 0.0.3

**Patch**

- Refactor templates to pass `sql` property instead of `text` property to
  `QueryInterface#query` method.
- Alias the `get_lock` call so it's easier to obtain the lock query result.


lib/index.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Assert = require("assert");
const Path = require("path");
const index_1 = require("./index");
describe('flock-mysql', function () {
    const dap = new index_1.DataAccessProvider({
        connectionOptions: {
            database: process.env.MYSQL_DATABASE,
            user: process.env.MYSQL_USER,
            password: process.env.MYSQL_PASSWORD
        }
    });
    let da = null;
    let qi = null;
    beforeEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            da = yield dap.provide();
            qi = da['qi']; // the QueryInterface
        });
    });
    afterEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            if (qi) {
                qi.query({
                    sql: `DROP TABLE IF EXISTS ${da.migrationTableName}`
                });
            }
            if (da) {
                yield da.close();
            }
        });
    });
    describe('TemplateProvider', function () {
        it('should provide a template file name when given a migration type that matches a template name', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const tp = new index_1.TemplateProvider();
                let fileName = yield tp.provideFileName('create-table');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/create-table.ejs'));
                fileName = yield tp.provideFileName('alter-table');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/alter-table.ejs'));
                fileName = yield tp.provideFileName('other');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/other.ejs'));
            });
        });
        it('should reject when given a migration type that does not match a template name', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const tp = new index_1.TemplateProvider();
                try {
                    yield tp.provideFileName('nope');
                }
                catch (error) {
                    Assert.strictEqual(error.code, 'UNSUPPORTED_MIGRATION_TYPE');
                }
            });
        });
    });
    describe('DataAccessProvider#provide', function () {
        it('should connect to the DB and acquire application lock', function () {
            return __awaiter(this, void 0, void 0, function* () {
                Assert.strictEqual(dap.migrationTableName, 'migration');
            });
        });
    });
    describe('PgDataAccess', function () {
        describe('#getMigratedMigrations', function () {
            it('should retrieve migrated migrations', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield qi.query({
                        sql: `CREATE TABLE IF NOT EXISTS \`${da.migrationTableName}\` (
              id varchar(512),
              created_at datetime DEFAULT NOW(),
              PRIMARY KEY(id)
            )`
                    });
                    yield qi.query({
                        sql: `INSERT INTO \`${da.migrationTableName}\` (id) VALUES(?)`,
                        values: ['one']
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.deepStrictEqual(migrated.map(x => x.id), ['one']);
                    Assert.ok(migrated[0].migratedAt instanceof Date);
                });
            });
        });
        describe('#migrate', function () {
            it('should migrate a migration', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield da.migrate('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.deepStrictEqual(migrated.map(x => x.id), ['two']);
                    Assert.ok(migrated[0].migratedAt instanceof Date);
                });
            });
        });
        describe('#rollback', function () {
            it('should rollback a migration', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield da.migrate('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    yield da.rollback('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.strictEqual(migrated.length, 0);
                });
            });
        });
    });
});


src/index.test.ts
import * as Assert from 'assert'
import * as Path from 'path'
import { DataAccessProvider, MySqlDataAccess, MySqlQueryInterface, TemplateProvider } from './index'

describe('flock-mysql', function () {
  const dap = new DataAccessProvider({
    connectionOptions: {
      database: process.env.MYSQL_DATABASE,
      user: process.env.MYSQL_USER,
      password: process.env.MYSQL_PASSWORD
    }
  })
  let da: MySqlDataAccess = null
  let qi: MySqlQueryInterface = null

  beforeEach(async function () {
    da = await dap.provide()
    qi = da['qi'] // the QueryInterface
  })

  afterEach(async function () {
    if (qi) {
      qi.query({
        sql: `DROP TABLE IF EXISTS ${da.migrationTableName}`
      })
    }
    if (da) {
      await da.close()
    }
  })

  describe('TemplateProvider', function () {
    it('should provide a template file name when given a migration type that matches a template name', async function () {
      const tp = new TemplateProvider()
      let fileName = await tp.provideFileName('create-table')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/create-table.ejs'))
      fileName = await tp.provideFileName('alter-table')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/alter-table.ejs'))
      fileName = await tp.provideFileName('other')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/other.ejs'))
    })

    it('should reject when given a migration type that does not match a template name', async function () {
      const tp = new TemplateProvider()
      try {
        await tp.provideFileName('nope')
      } catch (error) {
        Assert.strictEqual(error.code, 'UNSUPPORTED_MIGRATION_TYPE')
      }
    })
  })

  describe('DataAccessProvider#provide', function () {
    it('should connect to the DB and acquire application lock', async function () {
      Assert.strictEqual(dap.migrationTableName, 'migration')
    })
  })

  describe('PgDataAccess', function () {
    describe('#getMigratedMigrations', function () {
      it('should retrieve migrated migrations', async function () {
        await qi.query({
          sql:
            `CREATE TABLE IF NOT EXISTS \`${da.migrationTableName}\` (
              id varchar(512),
              created_at datetime DEFAULT NOW(),
              PRIMARY KEY(id)
            )`
        })
        await qi.query({
          sql: `INSERT INTO \`${da.migrationTableName}\` (id) VALUES(?)`,
          values: [ 'one' ]
        })
        const migrated = await da.getMigratedMigrations()
        Assert.deepStrictEqual(migrated.map(x => x.id), [ 'one' ])
        Assert.ok(migrated[0].migratedAt instanceof Date)
      })
    })

    describe('#migrate', function () {
      it('should migrate a migration', async function () {
        await da.migrate('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        const migrated = await da.getMigratedMigrations()
        Assert.deepStrictEqual(migrated.map(x => x.id), [ 'two' ])
        Assert.ok(migrated[0].migratedAt instanceof Date)
      })
    })

    describe('#rollback', function () {
      it('should rollback a migration', async function () {
        await da.migrate('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        await da.rollback('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        const migrated = await da.getMigratedMigrations()
        Assert.strictEqual(migrated.length, 0)
      })
    })
  })
})


src/index.ts
import * as Path from 'path'
import * as MySql from 'mysql'
import * as Flock from '@launchfort/flock'

export class TemplateProvider implements Flock.TemplateProvider {
  readonly migrationTypes = [ 'create-table', 'alter-table', 'other' ]

  provideFileName (migrationType: string) {
    if (this.migrationTypes.indexOf(migrationType) >= 0) {
      return Promise.resolve(Path.join(__dirname, 'templates', migrationType + '.ejs'))
    } else {
      return Promise.reject(Object.assign(
        new Error(`Unsupported migration type [${migrationType}]`),
        { code: 'UNSUPPORTED_MIGRATION_TYPE' }
      ))
    }
  }
}

export class DataAccessProvider implements Flock.DataAccessProvider {
  readonly migrationTableName: string
  readonly acquireLock: boolean
  readonly connectionOptions: MySql.ConnectionConfig

  constructor (options: { migrationTableName?: string, acquireLock?: boolean, connectionOptions: MySql.ConnectionConfig }) {
    const { migrationTableName = 'migration', acquireLock = true, connectionOptions } = options
    this.migrationTableName = migrationTableName
    this.acquireLock = acquireLock
    this.connectionOptions = Object.assign({
      host: 'localhost',
      port: 3306,
      user: 'MySQL',
      charset: 'UTF8_GENERAL_CI',
      timezone: 'Z',
      connectionTimeout: 10000,
      insecureAuth: false,
      debug: false,
      trace: false
    }, connectionOptions)
  }

  async provide () {
    const lock = 'flock-mysql'
    let locked = false

    return new Promise<MySql.Connection>((resolve, reject) => {
      const client: MySql.Connection = MySql.createConnection(this.connectionOptions)
      client.connect(error => {
        error ? reject(error) : resolve(client)
      })
    }).then(async (client) => {
      if (this.acquireLock) {
        const advisoryLockResult = await new Promise((resolve, reject) => {
          client.query(`SELECT GET_LOCK('${lock}', 2000) as locked`, (error, result) => {
            error ? reject(error) : resolve(result)
          })
        })
        locked = advisoryLockResult[0].locked === 1

        if (!locked) {
          await new Promise((resolve, reject) => {
            client.end(error => error ? reject(error) : resolve())
          })
          throw new Error(`Advisory lock "${lock}" could not be acquired.`)
        }
      }

      return new MySqlDataAccess(client, this.migrationTableName, { lock: locked ? lock : null })
    })
  }
}

export class MySqlDataAccess implements Flock.DataAccess {
  private client: MySql.Connection
  private qi: MySqlQueryInterface
  readonly migrationTableName: string
  readonly lock: string

  constructor (client: MySql.Connection, migrationTableName: string, { lock = null } = {}) {
    this.client = client
    this.qi = new MySqlQueryInterface(client)
    this.migrationTableName = migrationTableName
    this.lock = lock
  }

  async getMigratedMigrations () {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)

    if (!migrationTableExists) {
      return []
    }

    const result = await this.qi.query({
      sql: `SELECT id, created_at FROM \`${this.migrationTableName}\``
    })
    return (result.rows || []).map(x => {
      return { id: x.id, migratedAt: x.created_at }
    })
  }

  async migrate (migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>) {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)
    const hasMigrated = migrationTableExists ? (await this.hasMigrated(migrationId)) : false

    if (hasMigrated) {
      return
    }

    await this.qi.query({ sql: 'BEGIN' })
    try {
      await this.qi.query({
        sql:
          `CREATE TABLE IF NOT EXISTS \`${this.migrationTableName}\` (
            id varchar(512),
            created_at datetime DEFAULT NOW(),
            PRIMARY KEY(id)
          )`
      })
      await action(this.qi)
      await this.qi.query({
        sql: `INSERT INTO \`${this.migrationTableName}\` (id) VALUES(?)`,
        values: [ migrationId ]
      })
      await this.qi.query({ sql: 'COMMIT' })
    } catch (error) {
      await this.qi.query({ sql: 'ROLLBACK' })
      throw error
    }
  }

  async rollback (migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>) {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)

    if (!migrationTableExists) {
      return
    }

    const hasMigrated = await this.hasMigrated(migrationId)

    if (!hasMigrated) {
      return
    }

    await this.qi.query({ sql: 'BEGIN' })
    try {
      await action(this.qi)
      await this.qi.query({
        sql: `DELETE FROM \`${this.migrationTableName}\` WHERE id = ?`,
        values: [ migrationId ]
      })
      await this.qi.query({ sql: 'COMMIT' })
    } catch (error) {
      await this.qi.query({ sql: 'ROLLBACK' })
      throw error
    }
  }

  async close () {
    if (this.lock) {
      await this.qi.query({ sql: `SELECT RELEASE_LOCK('${this.lock}')` })
    }
    return this.client.end()
  }

  private async hasMigrated (migrationId: string) {
    const result = await this.qi.query({
      sql: `SELECT id FROM \`${this.migrationTableName}\` WHERE id = ?`,
      values: [ migrationId ]
    })
    return result.rowCount === 1
  }
}

export class MySqlQueryInterface implements Flock.QueryInterface {
  client: MySql.Connection

  constructor (client) {
    this.client = client
  }

  query (queryObject: { sql: string, values?: any[], timeout?: number }): Promise<Flock.QueryResult> {
    return new Promise((resolve, reject) => {
      this.client.query(queryObject, (error, result) => {
        error ? reject(error) : resolve({
          rowCount: result.length,
          rows: [].slice.call(result)
        })
      })
    })
  }

  async tableExists (tableName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      sql:
        `SELECT table_name
        FROM   information_schema.tables
        WHERE  table_name = ?`,
      values: [ tableName ]
    })
    return result.rowCount === 1
  }

  async columnExists (tableName: string, columnName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      sql:
        `SELECT column_name
        FROM information_schema.columns
        WHERE table_name = ?
        and column_name = ?`,
      values: [ tableName, columnName ]
    })
    return result.rowCount === 1
  }

  async columnDataType (tableName: string, columnName: string): Promise<string|null> {
    return this.inspectColumn(tableName, columnName).then(col => {
      return col ? col.data_type : null
    })
  }

  private async inspectColumn (tableName: string, columnName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      sql:
        `SELECT *
        FROM   information_schema.columns
        WHERE  table_name = ?
        AND column_name = ?`,
      values: [ tableName, columnName ]
    })
    return result.rowCount === 1 ? result.rows[0] : null
  }
}


lib/index.d.ts
import * as MySql from 'mysql';
import * as Flock from '@launchfort/flock';
export declare class TemplateProvider implements Flock.TemplateProvider {
    readonly migrationTypes: string[];
    provideFileName(migrationType: string): Promise<string>;
}
export declare class DataAccessProvider implements Flock.DataAccessProvider {
    readonly migrationTableName: string;
    readonly acquireLock: boolean;
    readonly connectionOptions: MySql.ConnectionConfig;
    constructor(options: {
        migrationTableName?: string;
        acquireLock?: boolean;
        connectionOptions: MySql.ConnectionConfig;
    });
    provide(): Promise<MySqlDataAccess>;
}
export declare class MySqlDataAccess implements Flock.DataAccess {
    private client;
    private qi;
    readonly migrationTableName: string;
    readonly lock: string;
    constructor(client: MySql.Connection, migrationTableName: string, { lock }?: {
        lock?: any;
    });
    getMigratedMigrations(): Promise<{
        id: any;
        migratedAt: any;
    }[]>;
    migrate(migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>): Promise<void>;
    rollback(migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>): Promise<void>;
    close(): Promise<void>;
    private hasMigrated;
}
export declare class MySqlQueryInterface implements Flock.QueryInterface {
    client: MySql.Connection;
    constructor(client: any);
    query(queryObject: {
        sql: string;
        values?: any[];
        timeout?: number;
    }): Promise<Flock.QueryResult>;
    tableExists(tableName: string): Promise<boolean>;
    columnExists(tableName: string, columnName: string): Promise<boolean>;
    columnDataType(tableName: string, columnName: string): Promise<string | null>;
    private inspectColumn;
}


lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Path = require("path");
const MySql = require("mysql");
class TemplateProvider {
    constructor() {
        this.migrationTypes = ['create-table', 'alter-table', 'other'];
    }
    provideFileName(migrationType) {
        if (this.migrationTypes.indexOf(migrationType) >= 0) {
            return Promise.resolve(Path.join(__dirname, 'templates', migrationType + '.ejs'));
        }
        else {
            return Promise.reject(Object.assign(new Error(`Unsupported migration type [${migrationType}]`), { code: 'UNSUPPORTED_MIGRATION_TYPE' }));
        }
    }
}
exports.TemplateProvider = TemplateProvider;
class DataAccessProvider {
    constructor(options) {
        const { migrationTableName = 'migration', acquireLock = true, connectionOptions } = options;
        this.migrationTableName = migrationTableName;
        this.acquireLock = acquireLock;
        this.connectionOptions = Object.assign({
            host: 'localhost',
            port: 3306,
            user: 'MySQL',
            charset: 'UTF8_GENERAL_CI',
            timezone: 'Z',
            connectionTimeout: 10000,
            insecureAuth: false,
            debug: false,
            trace: false
        }, connectionOptions);
    }
    provide() {
        return __awaiter(this, void 0, void 0, function* () {
            const lock = 'flock-mysql';
            let locked = false;
            return new Promise((resolve, reject) => {
                const client = MySql.createConnection(this.connectionOptions);
                client.connect(error => {
                    error ? reject(error) : resolve(client);
                });
            }).then((client) => __awaiter(this, void 0, void 0, function* () {
                if (this.acquireLock) {
                    const advisoryLockResult = yield new Promise((resolve, reject) => {
                        client.query(`SELECT GET_LOCK('${lock}', 2000) as locked`, (error, result) => {
                            error ? reject(error) : resolve(result);
                        });
                    });
                    locked = advisoryLockResult[0].locked === 1;
                    if (!locked) {
                        yield new Promise((resolve, reject) => {
                            client.end(error => error ? reject(error) : resolve());
                        });
                        throw new Error(`Advisory lock "${lock}" could not be acquired.`);
                    }
                }
                return new MySqlDataAccess(client, this.migrationTableName, { lock: locked ? lock : null });
            }));
        });
    }
}
exports.DataAccessProvider = DataAccessProvider;
class MySqlDataAccess {
    constructor(client, migrationTableName, { lock = null } = {}) {
        this.client = client;
        this.qi = new MySqlQueryInterface(client);
        this.migrationTableName = migrationTableName;
        this.lock = lock;
    }
    getMigratedMigrations() {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            if (!migrationTableExists) {
                return [];
            }
            const result = yield this.qi.query({
                sql: `SELECT id, created_at FROM \`${this.migrationTableName}\``
            });
            return (result.rows || []).map(x => {
                return { id: x.id, migratedAt: x.created_at };
            });
        });
    }
    migrate(migrationId, action) {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            const hasMigrated = migrationTableExists ? (yield this.hasMigrated(migrationId)) : false;
            if (hasMigrated) {
                return;
            }
            yield this.qi.query({ sql: 'BEGIN' });
            try {
                yield this.qi.query({
                    sql: `CREATE TABLE IF NOT EXISTS \`${this.migrationTableName}\` (
            id varchar(512),
            created_at datetime DEFAULT NOW(),
            PRIMARY KEY(id)
          )`
                });
                yield action(this.qi);
                yield this.qi.query({
                    sql: `INSERT INTO \`${this.migrationTableName}\` (id) VALUES(?)`,
                    values: [migrationId]
                });
                yield this.qi.query({ sql: 'COMMIT' });
            }
            catch (error) {
                yield this.qi.query({ sql: 'ROLLBACK' });
                throw error;
            }
        });
    }
    rollback(migrationId, action) {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            if (!migrationTableExists) {
                return;
            }
            const hasMigrated = yield this.hasMigrated(migrationId);
            if (!hasMigrated) {
                return;
            }
            yield this.qi.query({ sql: 'BEGIN' });
            try {
                yield action(this.qi);
                yield this.qi.query({
                    sql: `DELETE FROM \`${this.migrationTableName}\` WHERE id = ?`,
                    values: [migrationId]
                });
                yield this.qi.query({ sql: 'COMMIT' });
            }
            catch (error) {
                yield this.qi.query({ sql: 'ROLLBACK' });
                throw error;
            }
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.lock) {
                yield this.qi.query({ sql: `SELECT RELEASE_LOCK('${this.lock}')` });
            }
            return this.client.end();
        });
    }
    hasMigrated(migrationId) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.qi.query({
                sql: `SELECT id FROM \`${this.migrationTableName}\` WHERE id = ?`,
                values: [migrationId]
            });
            return result.rowCount === 1;
        });
    }
}
exports.MySqlDataAccess = MySqlDataAccess;
class MySqlQueryInterface {
    constructor(client) {
        this.client = client;
    }
    query(queryObject) {
        return new Promise((resolve, reject) => {
            this.client.query(queryObject, (error, result) => {
                error ? reject(error) : resolve({
                    rowCount: result.length,
                    rows: [].slice.call(result)
                });
            });
        });
    }
    tableExists(tableName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                sql: `SELECT table_name
        FROM   information_schema.tables
        WHERE  table_name = ?`,
                values: [tableName]
            });
            return result.rowCount === 1;
        });
    }
    columnExists(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                sql: `SELECT column_name
        FROM information_schema.columns
        WHERE table_name = ?
        and column_name = ?`,
                values: [tableName, columnName]
            });
            return result.rowCount === 1;
        });
    }
    columnDataType(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inspectColumn(tableName, columnName).then(col => {
                return col ? col.data_type : null;
            });
        });
    }
    inspectColumn(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                sql: `SELECT *
        FROM   information_schema.columns
        WHERE  table_name = ?
        AND column_name = ?`,
                values: [tableName, columnName]
            });
            return result.rowCount === 1 ? result.rows[0] : null;
        });
    }
}
exports.MySqlQueryInterface = MySqlQueryInterface;
